### ISSUE TITLE
Error when using a factor variable in a cluster variable

### ISSUE THREAD

**Issue by NilsEnevoldsen** [sergiocorreia/reghdfe#86]
```stata
. version
version 14.2

. which reghdfe
/Users/nilsschool/Library/Application Support/Stata/ado/plus/r/reghdfe.ado
*! version 4.1.0 28feb2017

. 
. sysuse auto, clear
(1978 Automobile Data)

. 
. reg price weight i.rep78, vce(cl foreign) //Model

Linear regression                               Number of obs     =         69
                                                F(0, 1)           =          .
                                                Prob > F          =          .
                                                R-squared         =     0.3631
                                                Root MSE          =     2414.8

                                (Std. Err. adjusted for 2 clusters in foreign)
------------------------------------------------------------------------------
             |               Robust
       price |      Coef.   Std. Err.      t    P>|t|     [95% Conf. Interval]
-------------+----------------------------------------------------------------
      weight |   2.440292   .6021941     4.05   0.154     -5.21131    10.09189
             |
       rep78 |
          2  |   783.9009   152.8068     5.13   0.123    -1157.693    2725.495
          3  |   1379.115   440.8581     3.13   0.197    -4222.519    6980.749
          4  |   2068.267   1737.752     1.19   0.445    -20011.97     24148.5
          5  |   3245.272   157.0189    20.67   0.031     1250.159    5240.386
             |
       _cons |  -3000.405   1866.802    -1.61   0.354    -26720.37    20719.56
------------------------------------------------------------------------------

. reghdfe price weight, absorb(i.rep78) vce(cl foreign) //Works
(converged in 1 iterations)

HDFE Linear regression                            Number of obs   =         69
Absorbing 1 HDFE group                            F(   1,      1) =      16.42
Statistics robust to heteroskedasticity           Prob > F        =     0.1540
                                                  R-squared       =     0.3631
                                                  Adj R-squared   =     0.3125
                                                  Within R-sq.    =     0.3537
Number of clusters (foreign) =          2         Root MSE        =  2414.8156

                                (Std. Err. adjusted for 2 clusters in foreign)
------------------------------------------------------------------------------
             |               Robust
       price |      Coef.   Std. Err.      t    P>|t|     [95% Conf. Interval]
-------------+----------------------------------------------------------------
      weight |   2.440292   .6021941     4.05   0.154     -5.21131    10.09189
------------------------------------------------------------------------------

Absorbed degrees of freedom:
-----------------------------------------------------+
 Absorbed FE | Categories  - Redundant  = Num. Coefs |
-------------+---------------------------------------|
       rep78 |         5           0           5     |
-----------------------------------------------------+

. reghdfe price weight, absorb(i.rep78) vce(cl i.foreign) //Error
i.foreign invalid name
           st_isnumvar():  3500  invalid Stata variable name
          __fload_data():     -  function returned error
FixedEffects::estimate_dof():     -  function returned error
                 <istmt>:     -  function returned error
r(3500);
```

**Comment by sergiocorreia** [comment]
Here it's also a bit tricky to allow i. with cluster, as I don't see an advantage wrt just using the variable name, and would require adding code that deals with factor variables in a more extensive way (currently we just allow x#y#z interactions, but no i. or c. prefixes)

**Comment by NilsEnevoldsen** [comment]
The advantage is that if I have a varlist with factor variables that I am using elsewhere in my do-file, I expect to be able to pass it to `reghdfe` without problems.

The secondary/related advantage is that I like being explicit with factor variable prefixes (`i.`, `c.`, `o.`, `b.`) as a code quality assurance. For example, if I use `i.` everywhere for categorical variables, I know that they won't be used accidentally as continuous variables.

It seems odd to support `reghdfe y x1 x2 , absorb(country#year) vce(cluster country#year)` without supporting `reghdfe y x1 x2 , absorb(i.country#i.year) vce(cluster i.country#i.year)`. The `#` seems like an indication that `reghdfe` supports factor variables, but now I see that it supports a sort of pseudo-factor variable notation.

At the least, this should be detected and throw an useful error, but supporting `#` without `i.` seems very strange to me.

**Comment by sergiocorreia** [comment]
Of all the 3 cases, this is probably the easiest to fix, but not sure if it would be enough for your do files.

There are several reasons why I can't use `syntax varlist(fv)` and do factor variables everywhere:

1. In `absorb()` I cannot expand `i.turn##c.gear` into `i.turn i.turn#c.gear` because that creates an extra factor variable that messes up the demeaning (makes it slower and less stable than doing it jointly).
2. Similarly, in absorb I can't split `i.turn#c.(gear weight)` (for the same reasons).
3. On the depvar, Stata's syntax doesn't allow `43.turn` which is something I actually find useful.
4. On the regressors, syntax by default expands interactions, and more importantly omits some factors it considers base factors. However, if that is done before singletons are computed, the sample would be different, and thus the actual list of base factors would be different.
4. Similarly, on cluster, I need to be careful to only allow turn#trunk but not i.trunk#c.turn

In other words, we try to keep as close as possible to standard factor variables, but have to deviate for good reasons, which means writing our own code (in several parts I'm working on a fork of the code Mark Schaeffer uses for ivreg2).

**Comment by NilsEnevoldsen** [comment]
Interesting background. Thanks. Perhaps a more appropriate enhancement would just be to make better errors/warnings about this stuff. Maybe you can parse the arguments for problematic expressions, like those I have used, and briefly explain in an error which factor variable syntax is not supported.

**Comment by sergiocorreia** [comment]
Sounds like a good idea; and easier to do at least on the short term.

**Comment by sergiocorreia** [comment]
Fixed; although some corner cases might remain
